<a href="https://colab.research.google.com/github/maoramirez69/14MBID/blob/main/3_Modelado_clasificacion_regresion.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
df_2048 = pd.read_csv('D:\\2_ASW\OneDrive - Asesoftware S.A.S\\Escritorio\\reutilizacion_06_bioactivity_data_3class_pIC50_pubchem_fp_2048.csv')
df_2048.shape

(7049, 2049)

In [ ]:
X_2048 = df_2048.drop('pIC50', axis=1)
Y_2048 = df_2048.pIC50

In [ ]:
#dividir dataset en entrenamiento y prueba
X_train_2048, X_test_2048, Y_train_2048, Y_test_2048 = train_test_split(X_2048, Y_2048, test_size=0.2,random_state=10)
X_train_2048.shape, Y_train_2048.shape, X_test_2048.shape, Y_test_2048.shape

((5639, 2048), (5639,), (1410, 2048), (1410,))

In [ ]:
#Modelo de regresion RFR
import numpy as np
model_RFR = RandomForestRegressor(n_estimators=102, max_depth=32, random_state=0, criterion='squared_error',  min_samples_split=2) #0.6685 32
model_RFR.fit(X_train_2048, Y_train_2048)
r2 = model_RFR.score(X_test_2048, Y_test_2048)
r2

0.6685283811262956

In [ ]:
from sklearn.tree import DecisionTreeRegressor 
  
model_DTR_2048 = DecisionTreeRegressor(random_state=6, max_depth=24, max_leaf_nodes=100, min_samples_leaf=1, splitter='random')
model_DTR_2048.fit(X_train_2048, Y_train_2048)
predictions = model_DTR_2048.predict(X_test_2048)

#r2 = model1.score(X_test, Y_test)

score = model_DTR_2048.score(X_test_2048, Y_test_2048)
print(score)

0.47273251769887925


In [ ]:
from sklearn.ensemble import ExtraTreesRegressor
ETR=ExtraTreesRegressor(n_estimators=100, random_state=0)
ETR_model=ETR.fit(X_train_2048, Y_train_2048)

In [ ]:
Y_pred_2048 =ETR.predict(X_test_2048)

In [ ]:
#calcular media de Y_pred_2048
media = Y_pred_2048.mean()
media

media = 7

#calcular mediana
mediana = np.median(Y_pred_2048)
mediana

4.800087676913794

In [ ]:
Y_train_2048_1 = [1 if i >= mediana else 0 for i in Y_train_2048]

In [ ]:
Y_test_2048_1 = [1 if i >= mediana else 0 for i in Y_test_2048]

In [ ]:
#Importar el modelo
import lightgbm as ltb
#Importar las métricas del modelo 
from sklearn import metrics
#se define el modelo
LGBM_model= ltb.LGBMClassifier()

## LGBM

In [ ]:
#Ajuste del modelo con los datos de entrenamiento
LGBM_fit=LGBM_model.fit(X_train_2048,Y_train_2048_1)

In [ ]:
#Hacer predicciones con el modelo (Varible respuesta sólo 0 o 1)
LGBM_pred=LGBM_model.predict(X_test_2048)
print(LGBM_pred)

[0 0 1 ... 0 0 0]


In [ ]:
#Hacer predicciones con el modelo (Probabilidades)
LGBM_pred_prob=LGBM_model.predict_proba(X_test_2048)

In [ ]:
#Matriz de confusión
#Se va a contruir la matriz de confusión  para el modelo 
from sklearn.metrics import confusion_matrix
confusion_matrix_LGBM = confusion_matrix(Y_test_2048_1, LGBM_pred)
Conf_matrix_LGBM=print(confusion_matrix_LGBM)

cm_display = metrics.ConfusionMatrixDisplay(confusion_matrix = Conf_matrix_LGBM, display_labels=[0,1],)
#mimportar plot
import matplotlib.pyplot as plt

[[605  77]
 [143 585]]


In [ ]:
#Métricas del modelo 
from sklearn.metrics import classification_report
nombres=['0','1']
#Métricas del modelo
print(classification_report(Y_test_2048_1, LGBM_pred,target_names=nombres))

              precision    recall  f1-score   support

           0       0.81      0.89      0.85       682
           1       0.88      0.80      0.84       728

    accuracy                           0.84      1410
   macro avg       0.85      0.85      0.84      1410
weighted avg       0.85      0.84      0.84      1410



In [ ]:
#Curva ROC del modelo
from sklearn import svm, datasets
from sklearn.metrics import roc_curve, auc
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import label_binarize
from sklearn.multiclass import OneVsRestClassifier
from sklearn.metrics import roc_auc_score

#Definir la función
def plot_roc_curve(fper, tper):
    plt.plot(fper, tper, color='blue', label='ROC')
    plt.plot([0, 1], [0, 1], color='green', linestyle='--')
    plt.xlabel('Tasa Falsos Positivos')
    plt.ylabel('Tasa Verdaderos Positivos')
    plt.title('Curva característica de funcionamiento del receptor (ROC)')
    plt.legend()
    plt.show()

#Curva ROC sin reducción de dimensionalidad
fper, tper, thresholds = roc_curve(Y_test_2048_1,LGBM_pred)
plot_roc_curve(fper, tper)
#mostrar el área bajo la curva
auc_score = roc_auc_score(Y_test_2048_1,LGBM_pred)
print(auc_score)

0.8453341013824885


## **Modelo XGBClassifier**

In [ ]:
#Importar el modelo de clasificación
from xgboost import XGBClassifier

In [ ]:
#Ajustar el modelo de clasificación
xgbc = XGBClassifier()

In [ ]:
#Ajustar el modelo de clasificación
xgbc = XGBClassifier()
XGBC_fit=xgbc.fit(X_train_2048,Y_train_2048_1)

XGBC_pred=xgbc.predict(X_test_2048)
print(XGBC_pred)
len(XGBC_pred)

[0 0 1 ... 0 0 0]


1410

In [ ]:
#Matriz de confusión
#Se va a contruir la matriz de confusión  para el modelo 
from sklearn.metrics import confusion_matrix
confusion_matrix_XGCB= confusion_matrix(Y_test_2048_1,XGBC_pred)
Conf_matrix_XGCB=print(confusion_matrix_XGCB)

[[611  71]
 [149 579]]


In [ ]:
#Métricas del modelo 
from sklearn.metrics import classification_report
nombres=['0','1']
#Métricas del modelo
print(classification_report(Y_test_2048_1, XGBC_pred,target_names=nombres))

              precision    recall  f1-score   support

           0       0.80      0.90      0.85       682
           1       0.89      0.80      0.84       728

    accuracy                           0.84      1410
   macro avg       0.85      0.85      0.84      1410
weighted avg       0.85      0.84      0.84      1410



In [ ]:
#calcular la precisión del modelo
from sklearn.metrics import accuracy_score
accuracy_score(Y_test_2048_1,XGBC_pred)

0.8439716312056738

## Regresión Logística

In [ ]:
from sklearn.linear_model import LogisticRegression

In [ ]:
#aplicar el modelo de Regresión Logística
LR=LogisticRegression()
LR_fit=LR.fit(X_train_2048,Y_train_2048_1)

LR_pred=LR.predict(X_test_2048)
print(LR_pred)

c:\Python3\lib\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


[0 0 1 ... 0 0 0]


In [ ]:
Model_clasification=LogisticRegression(random_state=0).fit(X_train_2048,Y_train_2048_1)

c:\Python3\lib\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [ ]:
#Información del modelo
print("Intercepto:", Model_clasification.intercept_)
print("Accuracy de entrenamiento:", Model_clasification.score(X_train_2048,Y_train_2048_1))

Intercepto: [-2.4362289]
Accuracy de entrenamiento: 0.9381095938996276


Se puede apreciar que en el modelo, el valor del intercepto $\beta_{0}=-2.4362289$. Por otroa lado, el modelo tiene un valor de accuracy de entrenamiento. En este caso, el accuracy es $$Accuracy=0.9381$$

Recordemos que, 
$$Accuracy=\frac{VP+VN}{VP+FP+VN+FN}$$

Por lo tanto, el modelo clasifica correctamente  a los datos de entrenamiento en un $93,20\%$.Si bien, la métrica de Accuracy es buena, se considera relevante buscar modelos que  tengan un mayor porcentaje de clasificación con los datos de entrenamiento.

In [ ]:
#calcular métricas del modelo
from sklearn.metrics import classification_report
nombres=['0','1']
#Métricas del modelo
print(classification_report(Y_test_2048_1, LR_pred,target_names=nombres))

              precision    recall  f1-score   support

           0       0.79      0.86      0.83       682
           1       0.86      0.79      0.82       728

    accuracy                           0.82      1410
   macro avg       0.83      0.82      0.82      1410
weighted avg       0.83      0.82      0.82      1410

